### PR TITLE
The pattern Role object

### PR DESCRIPTION
**Pull request by besok**
The pattern Role object

Based on https://github.com/iluwatar/java-design-patterns/issues/862

### CODE REVIEW

File: role-object/src/main/java/com/iluwatar/roleobject/ApplicationRoleObject.java [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+ * to the investment department’s view of customers.
+ * A client like the loan application may either work with objects of the CustomerCore class, using the interface class
+ * Customer, or with objects of concrete CustomerRole subclasses. Suppose the loan application knows a particular
+ * Customer instance through its Customer interface. The loan application may want to check whether the Customer
+ * object plays the role of Borrower.
+ * To this end it calls hasRole() with a suitable role specification. For the purpose of
+ * our example, let’s assume we can name roles with enum.
+ * If the Customer object can play the role named “Borrower,” the loan application will ask it to return a reference to the corresponding object.
+ * The loan application may now use this reference to call Borrower-specific operations.
+ */
```
Comment (iluwatar) [inline role-object/src/main/java/com/iluwatar/roleobject/ApplicationRoleObject.java]: JavaDoc @link could be used in the text to point to classes and methods

File: role-object/src/main/java/com/iluwatar/roleobject/ApplicationRoleObject.java [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+public class ApplicationRoleObject {
+
+    private static final Logger logger = LoggerFactory.getLogger(Role.class);
+
+    public static void main(String[] args) {
+        Customer customer = Customer.newCustomer(Borrower, Investor);
```
Comment (iluwatar) [inline role-object/src/main/java/com/iluwatar/roleobject/ApplicationRoleObject.java]: Local variable type inference (var) could be used. Check this throughout the code.

File: role-object/src/main/java/com/iluwatar/roleobject/ApplicationRoleObject.java (hunk 1 above)
Comment (iluwatar) [reply role-object/src/main/java/com/iluwatar/roleobject/ApplicationRoleObject.java]: Check other JavaDocs for this too.

File: role-object/src/main/java/com/iluwatar/roleobject/BorrowerRole.java [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+        return name;
+    }
+
+    public void setName(String name) {
+        this.name = name;
+    }
+
+    public String borrow(){
+        return String.join(" ",
+                "A borrower",name,"wants to get some money.");
```
Comment (iluwatar) [inline role-object/src/main/java/com/iluwatar/roleobject/BorrowerRole.java]: String.format might look cleaner

File: role-object/src/main/java/com/iluwatar/roleobject/ApplicationRoleObject.java (hunk 1 above)
Comment (besok) [reply role-object/src/main/java/com/iluwatar/roleobject/ApplicationRoleObject.java]: Yes, I agree. I took it from the description of patern

File: role-object/src/main/java/com/iluwatar/roleobject/ApplicationRoleObject.java (hunk 2 above)
Comment (besok) [reply role-object/src/main/java/com/iluwatar/roleobject/ApplicationRoleObject.java]: I agree, but i used java 8. If it is necessary of course I will update my java and change it to var :)

File: role-object/src/main/java/com/iluwatar/roleobject/BorrowerRole.java (hunk 3 above)
Comment (besok) [reply role-object/src/main/java/com/iluwatar/roleobject/BorrowerRole.java]: ok

Comment (besok) [reply role-object/src/main/java/com/iluwatar/roleobject/BorrowerRole.java]: done

File: role-object/src/main/java/com/iluwatar/roleobject/ApplicationRoleObject.java (hunk 1 above)
Comment (besok) [reply role-object/src/main/java/com/iluwatar/roleobject/ApplicationRoleObject.java]: done

File: role-object/src/main/java/com/iluwatar/roleobject/ApplicationRoleObject.java (hunk 2 above)
Comment (iluwatar) [reply role-object/src/main/java/com/iluwatar/roleobject/ApplicationRoleObject.java]: It's ok, I'll add this pattern to https://github.com/iluwatar/java-design-patterns/issues/984 to be updated later

**Comment by iluwatar** [comment]
Thanks @besok for this new pattern! :relaxed:

**Comment by besok** [comment]
No problem. I am pleased to help. :)

**Pull request merged**